src/node_contextify.cc

```diff
@@ -1391,6 +1395,300 @@ void MicrotaskQueueWrap::RegisterExternalReferences(
+    SetProtoMethod(isolate, tmpl, "runInCallbackScope", RunInCallbackScope);
+    SetProtoMethod(isolate, tmpl, "tryCloseAllHandles", TryCloseAllHandles);
+
+    env->set_localworker_constructor_template(tmpl);
+  }
+  return tmpl;
+}
+
+void LocalWorker::Initialize(Environment* env,
+                                   v8::Local<v8::Object> target) {
+  SetConstructorFunction(env->context(),
+                         target,
+                         "LocalWorker",
+                         GetConstructorTemplate(env),
+                         SetConstructorFunctionFlag::NONE);
+}
+
+void LocalWorker::RegisterExternalReferences(
+    ExternalReferenceRegistry* registry) {
+  registry->Register(New);
+  registry->Register(Start);
+  registry->Register(Load);
+  registry->Register(SignalStop);
+  registry->Register(Stop);
+  registry->Register(RunInCallbackScope);
+  registry->Register(TryCloseAllHandles);
+}
+
+LocalWorker::LocalWorkerScope::LocalWorkerScope(
+    LocalWorker* w)
+    : EscapableHandleScope(w->isolate_),
+      Scope(w->context()),
+      Isolate::SafeForTerminationScope(w->isolate_),
+      w_(w),
+      orig_can_be_terminated_(w->can_be_terminated_) {
+  w_->can_be_terminated_ = true;
+}
+
+LocalWorker::LocalWorkerScope::~LocalWorkerScope() {
+  w_->can_be_terminated_ = orig_can_be_terminated_;
+}
+
+Local<Context> LocalWorker::context() const {
+  return context_.Get(isolate_);
+}
+
+LocalWorker::LocalWorker(Environment* env, Local<Object> object)
+    : isolate_(env->isolate()), wrap_(env->isolate(), object) {
+  AddEnvironmentCleanupHook(env->isolate(), CleanupHook, this);
+  object->SetAlignedPointerInInternalField(0, this);
+
+  Local<Context> outer_context = env->context();
+  outer_context_.Reset(env->isolate(), outer_context);
+}
+
+LocalWorker* LocalWorker::Unwrap(
+    const FunctionCallbackInfo<Value>& args) {
+  Local<Value> value = args.This();
+  if (!value->IsObject() || value.As<Object>()->InternalFieldCount() < 1) {
+    THROW_ERR_INVALID_THIS(Environment::GetCurrent(args.GetIsolate()));
+    return nullptr;
+  }
+  return static_cast<LocalWorker*>(
+      value.As<Object>()->GetAlignedPointerFromInternalField(0));
+}
+
+void LocalWorker::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  new LocalWorker(env, args.This());
+}
+
+void LocalWorker::Start(const FunctionCallbackInfo<Value>& args) {
+  LocalWorker* self = Unwrap(args);
+  if (self == nullptr) return;
+  self->Start();
+}
+
+void LocalWorker::TryCloseAllHandles(
+    const FunctionCallbackInfo<Value>& args) {
+  auto count = 0;
+  LocalWorker* self = Unwrap(args);
+  self->env_->async_hooks()->clear_async_id_stack();
+  if (self->env_ != nullptr) {
+    auto env = self->env_;
+    for (auto w : *env->handle_wrap_queue()) {
+      count++;
+      w->Close();
+    }
+
+    for (auto w : *env->req_wrap_queue()) {
+      count++;
+      w->Cancel();
+    }
+  }
```

Comment:
@legendecas @joyeecheung These two loops are what you need to shut down any outstanding asynchronous work in the inner Node.js environment. We might need something similar for ShadowRealm too in case I/O operations are scheduled within it.